### PR TITLE
feat: precise fee, token_owner in offer datatype, handle unwraps 

### PR DESCRIPTION
**Pull request by ozwaldorf**
- 2 decimal precision fee, now 10.00% fee = `1000:nat`
- added `token_owner` in offer type
- `payment_address` is now `seller` in listing type
- `payment_address` is now `buyer` in offer type
- handle unwraps gracefully across the backend

## How?

- Done A (replace with a breakdown of the steps)
- Done B
- Done C

## Tickets?

- [Ticket 1](the-ticket-url-here)
- [Ticket 2](the-ticket-url-here)
- [Ticket 3](the-ticket-url-here)

## Contribution checklist?

- [ ] The commit messages are detailed
- [ ] It does not break existing features (unless required)
- [ ] I have performed a self-review of my own code
- [ ] Documentation has been updated to reflect the changes
- [ ] Tests have been added or updated to reflect the changes
- [ ] All code formatting pass
- [ ] All lints pass
- [ ] All tests pass

## Security checklist?

- [ ] Injection has been prevented (parameterized queries, no eval or system calls)
- [ ] Sensitive data has been identified and is being protected properly

## Demo?

Optionally, provide any screenshot, gif or small video.

### CODE REVIEW

File: marketplace/src/main.rs [hunk 1]
```diff
@@ -471,16 +499,15 @@ pub async fn direct_buy(nft_canister_id: Principal, token_id: Nat) -> MPApiResul
...
-        return Err(MPApiError::Other("Error calling balanceOf".to_string()));
-    } else if balance.ok().unwrap().clone() < listing.price.clone() {
+    if balance.clone() < listing.price.clone() {
         return Err(MPApiError::InsufficientFungibleBalance);
     }
 
+    // divide by 10000 to allow 2 digits of precision in the fee
     let owner_fee: Nat =
-        listing.price.clone() * collection.owner_fee_percentage.clone() / Nat::from(100);
+        listing.price.clone() * collection.owner_fee_percentage.clone() / Nat::from(10000);
```
Comment (punkbit) [inline marketplace/src/main.rs:510]: As we do the same computation in ln 565, let's have this as a utility method and use it accordingly, were required, please!

File: marketplace/src/main.rs [hunk 2]
```diff
@@ -697,18 +721,14 @@ pub async fn accept_offer(
...
-        if has_balance < offer_price.clone() {
-            return Err(MPApiError::InsufficientFungibleBalance);
-        }
+    if balance.clone() < offer_price.clone() {
+        return Err(MPApiError::InsufficientFungibleBalance);
     }
 
     let owner_fee: Nat =
-    offer_price.clone() * collection.owner_fee_percentage.clone() / Nat::from(100);
+    offer_price.clone() * collection.owner_fee_percentage.clone() / Nat::from(10000);
```
Comment (punkbit) [inline marketplace/src/main.rs:731]: Related to comment on reusable utility fn, see above.

File: marketplace/src/main.rs [hunk 3]
```diff
@@ -533,8 +560,9 @@ pub async fn direct_buy(nft_canister_id: Principal, token_id: Nat) -> MPApiResul
         return Err(MPApiError::TransferNonFungibleError);
     }
 
+    // divide by 10000 to allow 2 digits of precision in the fee
     let owner_fee: Nat =
-        listing.price.clone() * collection.owner_fee_percentage.clone() / Nat::from(100);
+        listing.price.clone() * collection.owner_fee_percentage.clone() / Nat::from(10000);
```
Comment (punkbit) [inline marketplace/src/main.rs:565]: Related to comment on reusable utility fn, see above.

**Review by punkbit — CHANGES_REQUESTED** [review]

**Comment by ozwaldorf** [comment]
will handle calc_fee() utility method in new PR, along with multiple fee fields

**Review by punkbit — APPROVED** [review]

**Pull request merged**